### PR TITLE
PI-1706: Log when CheckoutTheme values are changed

### PR DESCRIPTION
**Pull request by chintan-soni-cko**
## Proposed changes

1. Added log events for Checkout Theme value changed
2. Added Unit test to support log events' changes

## Types of changes

What types of changes does your code introduce to frames-android?
_Put an `x` in the boxes that apply_

* [ ] Bugfix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [X] Lint and unit tests pass locally with my changes
* [X] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)

## Further comments
1. Required to work out for sending font details like font name to events.
2. Just to update, Currently changing the status bar colour is not supported. Included all attributes for customization except font details. Additionally, sending extra details for textcolorPrimary, colorControlActivated to events if the merchant has added this value or it would be sending the default colours.

### CODE REVIEW

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    init {
+        colorPrimary = geThemeColorResource(R.attr.colorPrimary)
+        colorAccent = geThemeColorResource(R.attr.colorAccent)
+        colorButtonNormal = geThemeColorResource(R.attr.colorButtonNormal)
+        colorControlNormal = geThemeColorResource(R.attr.colorControlNormal)
+        textColorPrimary = geThemeColorResource(android.R.attr.textColorPrimary)
+        colorControlActivated = geThemeColorResource(R.attr.colorControlActivated)
+    }
+
+    @SuppressLint("RestrictedApi")
```
Comment (deepesh-vasthimal-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:27]: Question : Any reason to suppress the lint here ?

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    internal fun geThemeColorResource(colorAttribute: Int): MutableMap<String, Any?> {
```
Comment (deepesh-vasthimal-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:28]: can we add some unit tests for geThemeColorResource method.

File: android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt [hunk 3]
```diff
@@ -1,18 +1,28 @@
...
+import com.checkout.android_sdk.Utils.CheckoutTheme
 import com.checkout.android_sdk.Utils.Environment
+import com.checkout.android_sdk.Utils.filterNotNullValues
 import com.checkout.eventlogger.domain.model.MonitoringLevel
 import java.util.Locale
 
 object FramesLoggingEventDataProvider {
 
-    fun logPaymentFormPresentedEvent(): FramesLoggingEvent {
+    fun logPaymentFormPresentedEvent(context: Context): FramesLoggingEvent {
```
Comment (deepesh-vasthimal-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt:11]: Can you not just pass CheckoutTheme into logPaymentFormPresentedEvent as param instead of Context ?

File: android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt [hunk 4]
```diff
@@ -1,36 +1,58 @@
...
+            FramesLoggingEventDataProvider.logPaymentFormPresentedEvent(mockContext).monitoringLevel)
 
         assertEquals(getExpectedPaymentFormPresentedLoggingEvent().properties,
-            FramesLoggingEventDataProvider.logPaymentFormPresentedEvent().properties)
+            FramesLoggingEventDataProvider.logPaymentFormPresentedEvent(mockContext).properties)
 
     }
 
 
     private fun getExpectedPaymentFormPresentedLoggingEvent(): FramesLoggingEvent {
```
Comment (deepesh-vasthimal-cko) [inline android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt:41]: What is the use of this test helper function when if looks exact similar to your core logic of logPaymentFormPresentedEvent ?

File: android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt [hunk 5]
```diff
@@ -1,36 +1,58 @@
...
+            PaymentFormLanguageEventAttribute.colorButtonNormal to checkoutTheme.getColorButtonNormalProperty(),
+            PaymentFormLanguageEventAttribute.colorControlNormal to checkoutTheme.getColorButtonNormalProperty(),
+            PaymentFormLanguageEventAttribute.textColorPrimary to checkoutTheme.getTextColorPrimaryProperty(),
+            PaymentFormLanguageEventAttribute.colorControlActivated to checkoutTheme.getColorControlActivatedProperty(),
         )
         return FramesLoggingEvent(
             MonitoringLevel.INFO,
             FramesLoggingEventType.PAYMENT_FORM_PRESENTED,
-            properties = eventData
+            properties = eventData.filterNotNullValues()
```
Comment (deepesh-vasthimal-cko) [inline android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt:55]: You should hardcode the expected values here and try to add some Null values here to add more test cases ( both null and not null to test behaviour ) instead of calling filterNotNullValues on eventData.

**Review by deepesh-vasthimal-cko — CHANGES_REQUESTED** [review]
Few comments otherwise looks good. Please have a look.

File: android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt (hunk 3 above)
Comment (chintan-soni-cko) [reply android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt:11]: Yes, we can pass the CheckoutTheme as well.

File: android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt (hunk 4 above)
Comment (chintan-soni-cko) [reply android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt:41]: So, we are mapping properties of events like theme and local device language with production environment to method name as getExpectedPaymentFormPresentedLoggingEvent's property which is created using mocking context to validate the production environment's properties.

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt (hunk 1 above)
Comment (chintan-soni-cko) [reply android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:27]: I was using getThemeColorAttr method (utility method for getting theme colours from attributes)of the androidx library, which may be subject to change in future releases of androidx versions. The reason for using supresslint is to avoid conflict of API change in future for this specific method for the safe call. However, I have removed that and applied the core native method to get the colours. I will keep a note to not use as much as public API methods which aim to change in the future.

File: android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt (hunk 4 above)
Comment (deepesh-vasthimal-cko) [reply android-sdk/src/test/java/com/checkout/android_sdk/logging/FramesLoggingEventsTest.kt:42]: Can we avoid writing additional code in test to make the test work ? instead use the logic within the SDK as its repeating in logPaymentFormPresentedEvent.

**Review by deepesh-vasthimal-cko — APPROVED** [review]
Just one last comment otherwise looks good.

File: android-sdk/src/main/java/com/checkout/android_sdk/PaymentForm.java [hunk 6]
```diff
@@ -586,14 +587,14 @@ private synchronized FramesLogger getFramesLogger() {
      */
     private void sendPaymentFormPresentedEvent(FramesLogger framesLogger) {
         post(() -> {
-            logPaymentFormPresentedEvent(framesLogger);
+            logPaymentFormPresentedEvent(framesLogger, new CheckoutTheme(mContext));
```
Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/PaymentForm.java:590]: Should be any Context null-check be done here for the cases when module is used without UI?
As far as I can see CheckoutTheme works with non nullable context.

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
+package com.checkout.android_sdk.Utils
+
+import android.content.Context
+import android.content.res.Resources
+import android.util.TypedValue
+import com.checkout.android_sdk.R
+
+internal class CheckoutTheme(val context: Context) : CheckoutThemeInterface {
```
Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:8]: Is context field used outside of the class?
It would be nice to make it private if possible.

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutThemeInterface.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.checkout.android_sdk.Utils
+
+internal interface CheckoutThemeInterface {
+    fun getColorPrimaryProperty(): MutableMap<String, Any?>
+    fun getColorAccentProperty(): MutableMap<String, Any?>
+    fun getColorButtonNormalProperty(): MutableMap<String, Any?>
+    fun getTextColorPrimaryProperty(): MutableMap<String, Any?>
+    fun getColorControlActivatedProperty(): MutableMap<String, Any?>
+    fun getColorControlNormalProperty(): MutableMap<String, Any?>
```
Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutThemeInterface.kt:9]: From the first look these functions look more like getters.
If it's so, is there any need for map mutability in this case?

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        val typedValue = TypedValue()
+        val theme: Resources.Theme = context.theme
+        theme.resolveAttribute(colorAttribute, typedValue, true)
+        val color = typedValue.data
+
+        color.let {
+            colorValues["alpha"] = color shr 24 and 0xFF
+            colorValues["red"] = color shr 16 and 0xFF
+            colorValues["green"] = color shr 8 and 0xFF
+            colorValues["blue"] = color and 0xFF
```
Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:37]: There is a possibility to use `android.graphics.Color`.
Mainly that class provides the same functionality under the hood.
So at the end we can simplify that function a bit, f.e.:

```
    private fun geThemeColorResource(colorAttribute: Int): MutableMap<String, Any?> {
        val colorValues = mutableMapOf<String, Any?>()
        val typedValue = TypedValue()

        context.theme.resolveAttribute(colorAttribute, typedValue, true)

        typedValue.data.let { color ->
            "alpha" to Color.alpha(color)
            "red" to Color.red(color)
            "green" to Color.green(color)
            "blue" to Color.blue(color)
        }

        return colorValues
    }
```

Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:37]: There is a possibility to use `android.graphics.Color`.
Mainly that class provides the same functionality under the hood.
So at the end we can simplify that function a bit.

Comment (maxim-nosov-cko) [reply android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutTheme.kt:37]: As far as I'm concerned, `typedValue.data` can't be null, please correct me if I'm wrong.
If it's so, then `let {}` can be  removed. F.e:
```
    private fun geThemeColorResource(colorAttribute: Int): MutableMap<String, Any?> {
        val typedValue = TypedValue()
        context.theme.resolveAttribute(colorAttribute, typedValue, true)
        val color = typedValue.data

        return mutableMapOf(
            "alpha" to Color.alpha(color),
            "red" to Color.red(color),
            "green" to Color.green(color),
            "blue" to Color.blue(color)
        )
    }
```

File: android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt [hunk 10]
```diff
@@ -1,18 +1,26 @@
 package com.checkout.android_sdk.logging
+import com.checkout.android_sdk.Utils.CheckoutTheme
 import com.checkout.android_sdk.Utils.Environment
+import com.checkout.android_sdk.Utils.filterNotNullValues
 import com.checkout.eventlogger.domain.model.MonitoringLevel
-import java.util.Locale
+import java.util.*
```
Comment (maxim-nosov-cko) [inline android-sdk/src/main/java/com/checkout/android_sdk/logging/FramesLoggingEventDataProvider.kt:6]: It would be nice to avoid wildcard usage in imports when we need just a few classes.

File: android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutThemeInterface.kt (hunk 8 above)
Comment (chintan-soni-cko) [reply android-sdk/src/main/java/com/checkout/android_sdk/Utils/CheckoutThemeInterface.kt:9]: It's not required while obtaining data. Using the same data type while adding it. However, I will use Map for it.

**Review by maxim-nosov-cko — APPROVED** [review]
LGTM
Good job  👍
